src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp

```diff
@@ -77,7 +77,7 @@ class ShenandoahResetBitmapTask : public WorkerTask {
     ShenandoahHeap* heap = ShenandoahHeap::heap();
     ShenandoahMarkingContext* const ctx = heap->marking_context();
     while (region != nullptr) {
-      bool needs_reset = _generation->contains(region) || !region->is_affiliated();
+      bool needs_reset = !region->is_affiliated() || _generation->contains(region);
```

Comment:
We should really read the affiliation for the whole test, but will have to create a new method to achieve that to keep the code clean and encapsulated.

Reply:
Hmm, you reckon the affiliation is changing between first testing for `_generation->contains` and then testing for `!region->is_affiliated()`? I suppose this _could_ happen, but the fix feels fragile (not sure we can guarantee the compiler won't reorder the tests). We do have `ShenandoahHeapRegion::affiliation`. What about:
```C
  auto affiliation = region->affiliation();
  bool needs_reset = (affiliation == ShenandoahAffiliation::FREE || _generation->contains(affiliation))
```
Where `ShenandoahGeneration::contains(ShenandoaAffiliation affiliation)` tests if the affiliation and the generation type match?

Reply:
I didn't notice we have `ShenandoahGeneration::contains(ShenandoaAffiliation affiliation) `, I'll update the PR, thanks.

Reply:
I have added `ShenandoahGeneration::contains(ShenandoaAffiliation affiliation)` and updated the PR.